Harmony/Tools/AccessTools.cs

```diff
@@ -15,6 +15,11 @@ namespace HarmonyLib
 	///
 	public static class AccessTools
 	{
+		/// <summary>
+		/// A cache for the <see cref="ICollection{T}.Add"/> or similar Add methods for different types.
+		/// </summary>
+		private static readonly Dictionary<Type, FastInvokeHandler> handlerCache = new Dictionary<Type, FastInvokeHandler>();
```

Comment:
This isn't thread-safe. Ideally we'd use a ConcurrentDictionary, but that isn't supported on net35.

In other areas, we've been using locks as a workaround, but if performance is a concern, then there would have to be two code paths:

net35: uses Dictionary and either:
a) locks on that Dictionary upon usage
b) use ThreadStatic (not ideal for a cache) with manual per-thread initialization

Other: use ConcurrentDictionary

That said, I think always locking is fine, since the dict access performance is minor compared to the addInvoker usage.

Reply:
Good to hear that Harmony is moving to multi-thread

Reply:
I will switch the code to ConcurrentDictionary

Reply:
You can't use ConcurrentDictionary on .NET Framework 3.5 though, which Harmony must still support. This is why I suggest just locking instead.